File: gateway/http.py
"""
HTTP Gateway - HTTP API 网关
"""

from .base import Gateway
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from fastapi.middleware.cors import CORSMiddleware
import uvicorn
import asyncio


class HTTPGateway(Gateway):
    """HTTP API 网关"""
    
    def __init__(self, host: str = "0.0.0.0", port: int = 8080):
        super().__init__()
        self._host = host
        self._port = port
        self._app = FastAPI(title="XiaoMengCore API")
        self._server = None
        
        self._setup_routes()
    
    def _setup_routes(self):
        """设置路由"""
        self._app.add_middleware(
            CORSMiddleware,
            allow_origins=["*"],
            allow_credentials=True,
            allow_methods=["*"],
            allow_headers=["*"],
        )
        
        @self._app.post("/chat")
        async def chat(request: Request):
            data = await request.json()
            message = data.get("message", "")
            user_id = data.get("user_id", "anonymous")
            
            if self._message_handler:
                from models import Message, User, UserLevel, Source
                
                user = User(
                    user_id=user_id,
                    level=UserLevel.STRANGER,
                    source=Source.WEB
                )
                
                msg = Message(
                    content=message,
                    user=user,
                    source=Source.WEB
                )
                
                response = await self._message_handler(msg)
                
                return JSONResponse({
                    "success": True,
                    "response": response.content if response else "",
                    "emotion": response.emotion.value if response and response.emotion else "neutral"
                })
            
            return JSONResponse({"success": False, "error": "No handler"})
        
        @self._app.get("/health")
        async def health():
            return {"status": "ok"}
    
    async def start(self):
        """启动 HTTP 网关"""
        config = uvicorn.Config(
            self._app,
            host=self._host,
            port=self._port,
            log_level="warning"
        )
        self._server = uvicorn.Server(config)
        
        asyncio.create_task(self._server.serve())
    
    async def stop(self):
        """停止 HTTP 网关"""
        if self._server:
            self._server.should_exit = True
        self._running = False
    
    async def send(self, target: str, message: str):
        """发送消息（HTTP 不支持主动发送）"""
        pass

File: gateway_rust/src/http.rs
use axum::{
    extract::Json,
    routing::{get, post},
    Router,
};
use std::sync::Arc;
use tokio::sync::RwLock;
use tower_http::cors::{Any, CorsLayer};
use anyhow::Result;

use crate::models::{Message, User, UserLevel};
use crate::gateway::{BaseGateway, MessageHandler};

#[derive(Debug, Clone)]
pub struct HttpGatewayConfig {
    pub host: String,
    pub port: u16,
}

impl Default for HttpGatewayConfig {
    fn default() -> Self {
        Self {
            host: "0.0.0.0".to_string(),
            port: 8080,
        }
    }
}

pub struct HttpGateway {
    config: HttpGatewayConfig,
    running: Arc<RwLock<bool>>,
    message_handler: Arc<RwLock<Option<MessageHandler>>>,
}

impl HttpGateway {
    pub fn new(config: HttpGatewayConfig) -> Self {
        Self {
            config,
            running: Arc::new(RwLock::new(false)),
            message_handler: Arc::new(RwLock::new(None)),
        }
    }
}

#[async_trait::async_trait]
impl BaseGateway for HttpGateway {
    fn source(&self) -> crate::models::Source {
        crate::models::Source::Web
    }

    async fn start(&mut self) -> Result<()> {
        let addr = format!("{}:{}", self.config.host, self.config.port);
        tracing::info!("HTTP Gateway starting on {}", addr);

        *self.running.write().await = true;

        let message_handler = self.message_handler.clone();
        let running = self.running.clone();

        tokio::spawn(async move {
            let app = Router::new()
                .route("/api/chat", post({
                    let message_handler = message_handler.clone();
                    move |Json(req): Json<ChatRequest>| {
                        let message_handler = message_handler.clone();
                        async move {
                            let user = User::new(req.user_id, req.user_level);
                            let message = Message::web(user, req.content);
                            
                            let handler = message_handler.read().await;
                            if let Some(handler) = handler.as_ref() {
                                match handler(message).await {
                                    Ok(response) => Json(ChatResponse {
                                        success: true,
                                        response: Some(response.content),
                                        error: None,
                                    }),
                                    Err(e) => Json(ChatResponse {
                                        success: false,
                                        response: None,
                                        error: Some(e.to_string()),
                                    }),
                                }
                            } else {
                                Json(ChatResponse {
                                    success: false,
                                    response: None,
                                    error: Some("No handler".to_string()),
                                })
                            }
                        }
                    }
                }))
                .route("/api/health", get(|| async { "OK" }))
                .layer(CorsLayer::new().allow_origin(Any).allow_methods(Any).allow_headers(Any));

            let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();
            axum::serve(listener, app).await.ok();
        });

        Ok(())
    }

    async fn stop(&mut self) -> Result<()> {
        *self.running.write().await = false;
        tracing::info!("HTTP Gateway stopped");
        Ok(())
    }

    async fn send_message(&self, _response: crate::models::Response) -> Result<()> {
        Ok(())
    }

    fn is_running(&self) -> bool {
        futures::executor::block_on(async {
            *self.running.read().await
        })
    }

    fn set_message_handler(&mut self, handler: MessageHandler) {
        futures::executor::block_on(async {
            let mut h = self.message_handler.write().await;
            *h = Some(handler);
        });
    }
}

#[derive(serde::Deserialize)]
struct ChatRequest {
    content: String,
    user_id: String,
    #[serde(default)]
    user_level: UserLevel,
}

#[derive(serde::Serialize)]
struct ChatResponse {
    success: bool,
    response: Option<String>,
    error: Option<String>,
}
